### ISSUE TITLE
NullReferenceException using NTLM with CredentialCache.DefaultNetworkCredentials

### ISSUE THREAD

**Issue by marco-olimpio** [jstedfast/MailKit#1353]
Hi Jeffrey! 

**First of all, thanks a lot for your work with MailKit and MimeKit!**

I'm trying to use the MaillKit through a authenticated smtp server but I'm getting the "Object reference not set instance of object". I'm capable to receive the chanllenge from the server but it throws an excception after that. I do not know if you missed the any config.

**Describe the bug**
Object reference not set instance of object

 - OS: Client Windows 10 Enterprise / Server: Windows 2012 R2 / SMTP: Exchange 2013 CU23
 - .NET Runtime: .NET Framework 4.7.1
 - .NET Framework: .NET Framework 4.7.1
 - MailKit Version: 3.1.1
 - MimeKit Version: 3.1.1
 - NSspi Version: 0.3.1

**Exception**
```
System.NullReferenceException: La référence d'objet n'est pas définie à une instance d'un objet.
   à System.Security.Cryptography.HMAC.InitializeKey(Byte[] key)
   à System.Security.Cryptography.HMACMD5..ctor(Byte[] key)
   à MailKit.Security.Ntlm.NtlmUtils.HMACMD5(Byte[] key, Byte[][] values) dans D:\src\MailKit\MailKit\Security\Ntlm\NtlmUtils.cs:ligne 100
   à MailKit.Security.Ntlm.NtlmAuthenticateMessage.ComputeNtlmV2(String targetName, Boolean unverifiedTargetName, Byte[] channelBinding) dans D:\src\MailKit\MailKit\Security\Ntlm\NtlmAuthenticateMessage.cs:ligne 326
   à MailKit.Security.SaslMechanismNtlm.GetChallengeResponse(String domain, String userName, String password, Byte[] token, Int32 startIndex, Int32 length) dans D:\src\MailKit\MailKit\Security\SaslMechanismNtlm.cs:ligne 291
   à MailKit.Security.SaslMechanismNtlm.Challenge(Byte[] token, Int32 startIndex, Int32 length, CancellationToken cancellationToken) dans D:\src\MailKit\MailKit\Security\SaslMechanismNtlm.cs:ligne 276
   à MailKit.Security.SaslMechanism.Challenge(String token, CancellationToken cancellationToken) dans D:\src\MailKit\MailKit\Security\SaslMechanism.cs:ligne 359
   à MailKit.Net.Smtp.SmtpClient.<AuthenticateAsync>d__93.MoveNext() dans D:\src\MailKit\MailKit\Net\Smtp\SmtpClient.cs:ligne 934
--- Fin de la trace de la pile à partir de l'emplacement précédent au niveau duquel l'exception a été levée ---
   à System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   à System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   à MailKit.Net.Smtp.SmtpClient.Authenticate(SaslMechanism mechanism, CancellationToken cancellationToken) dans D:\src\MailKit\MailKit\Net\Smtp\SmtpClient.cs:ligne 1008
   à SendEmail.Program.EnvoiCourrielMailKit() dans C:\Users\rolm005\source\repos\WindowsApp1\SendEmail\Program.cs:ligne 202
```
**To Reproduce**
Simply execute the code.

**Expected behavior**
User authenticated and capable to send emails.

**Code Snippets**
```csharp
private static void EnvoiCourrielMailKit()
{
	//Message
	MimeMessage messageElectronique = new MimeMessage()
	{
		Sender = MailboxAddress.Parse("marco@courriel.com"),
		From = { MailboxAddress.Parse("marco@courriel.com") },
		Subject = "TEST - MailKit",
		Body = new TextPart(TextFormat.Html) { Text = "MARCO MailKit!" },
		To = { MailboxAddress.Parse("marco@courriel.com") }
	};

	using (SmtpClient client = new SmtpClient(new ProtocolLogger("SMTP-AUTH.log")))
	{
		try
		{
			client.ServerCertificateValidationCallback = (s, c, h, e) => true;
                        client.Connect(AuthMail, 25, SecureSocketOptions.StartTlsWhenAvailable);

			Console.WriteLine("Types of athentication mechanisms: ");
			foreach (var clientAuthenticationMechanism in client.AuthenticationMechanisms)
			{
				Console.WriteLine("    {0}", clientAuthenticationMechanism);
			}
			client.AuthenticationMechanisms.Clear();
			client.AuthenticationMechanisms.Add("NTLM");
			
			client.Authenticate(mechanism: new SaslMechanismNtlm(CredentialCache.DefaultNetworkCredentials));
			//var cred = SaslMechanismNtlmIntegrated.Create("NTLM", CredentialCache.DefaultNetworkCredentials);
			//client.Authenticate(cred);
			client.Send(messageElectronique);
		}
		catch (Exception e)
		{
			Console.WriteLine(e);
		}
		finally
		{
                        Console.WriteLine("End using NTLM - CredentialCache.DefaultNetworkCredentials");
			if (client.IsConnected)
				client.Disconnect(true);
		}
	}
}
```

**Protocol Logs**

```
Connected to smtp://smtp-auth.intranet:25/?starttls=when-available
S: 220 SERVER- Relais AUTH - Postes
C: EHLO MYMACHINE
S: 250-SERVER Hello [177.77.77.177]
S: 250-SIZE 52428800
S: 250-PIPELINING
S: 250-DSN
S: 250-ENHANCEDSTATUSCODES
S: 250-STARTTLS
S: 250-AUTH NTLM LOGIN
S: 250-8BITMIME
S: 250-BINARYMIME
S: 250 CHUNKING
C: STARTTLS
S: 220 2.0.0 SMTP server ready
C: EHLO MYMACHINE
S: 250-SERVER Hello [177.77.77.177]
S: 250-SIZE 52428800
S: 250-PIPELINING
S: 250-DSN
S: 250-ENHANCEDSTATUSCODES
S: 250-AUTH NTLM LOGIN
S: 250-8BITMIME
S: 250-BINARYMIME
S: 250 CHUNKING
C: AUTH NTLM ********
S: 334 [CHALLENGE - 260 CHARS]
```

**Additional context**
I do not know if It help but I can use the `Net.Mail.SmtpClient` with the `SmtpClient.UseDefaultCredentials = true` or the `CredentialCache.DefaultNetworkCredentials` to send authenticated emails.
**But problably I'm missing something**. The server responds the `client.AuthenticationMechanisms` with `LOGIN` and `NTLM`.
If you need the challenge I can send to your email.

Thanks a lot, Jeffrey!

**Comment by marco-olimpio** [comment]
I've just tested with the V3.2 but changed the line numbers of the stacktrace

```
System.NullReferenceException: La référence d'objet n'est pas définie à une instance d'un objet.
   à System.Security.Cryptography.HMAC.InitializeKey(Byte[] key)
   à System.Security.Cryptography.HMACMD5..ctor(Byte[] key)
   à MailKit.Security.Ntlm.NtlmUtils.HMACMD5(Byte[] key, Byte[][] values) dans D:\src\MailKit\MailKit\Security\Ntlm\NtlmUtils.cs:ligne 100
   à MailKit.Security.Ntlm.NtlmAuthenticateMessage.ComputeNtlmV2(String targetName, Boolean unverifiedTargetName, Byte[] channelBinding) dans D:\src\MailKit\MailKit\Security\Ntlm\NtlmAuthenticateMessage.cs:ligne 326
   à MailKit.Security.SaslMechanismNtlm.GetChallengeResponse(String domain, String userName, String password, Byte[] token, Int32 startIndex, Int32 length) dans D:\src\MailKit\MailKit\Security\SaslMechanismNtlm.cs:ligne 299
   à MailKit.Security.SaslMechanismNtlm.Challenge(Byte[] token, Int32 startIndex, Int32 length, CancellationToken cancellationToken) dans D:\src\MailKit\MailKit\Security\SaslMechanismNtlm.cs:ligne 284
   à MailKit.Security.SaslMechanism.Challenge(String token, CancellationToken cancellationToken) dans D:\src\MailKit\MailKit\Security\SaslMechanism.cs:ligne 359
   à MailKit.Net.Smtp.SmtpClient.<AuthenticateAsync>d__93.MoveNext() dans D:\src\MailKit\MailKit\Net\Smtp\SmtpClient.cs:ligne 934
--- Fin de la trace de la pile à partir de l'emplacement précédent au niveau duquel l'exception a été levée ---
   à System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   à System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   à MailKit.Net.Smtp.SmtpClient.Authenticate(SaslMechanism mechanism, CancellationToken cancellationToken) dans D:\src\MailKit\MailKit\Net\Smtp\SmtpClient.cs:ligne 1008
   à SendEmail.Program.EnvoiCourrielMailKit() dans C:\Users\USER\source\repos\WindowsApp1\SendEmail\Program.cs:ligne 209
```

**Comment by jstedfast** [comment]
The only way that I can see that this could happen is if the userName and password are null.

The NRE happens because the `ExportedSessionKey` is null here: https://github.com/jstedfast/MailKit/blob/master/MailKit/Security/Ntlm/NtlmAuthenticateMessage.cs#L327

And the only way that can be null is if the following code path is executed:

https://github.com/jstedfast/MailKit/blob/77eb92573cec5b38bafb3a1ebdd819693ac4162f/MailKit/Security/Ntlm/NtlmUtils.cs#L211-L215

**Comment by jstedfast** [comment]
Try this:

```csharp
var credentials = CredentialCache.DefaultNetworkCredentials;
Console.WriteLine ("UserName = '{0}'; Password = '{1}'", credentials.UserName, credentials.Password);
```

**Comment by marco-olimpio** [comment]
Hi Jeffrey, yeah. You're completely right! 
I've mixed the CredentialCache.DefaultNetworkCredentials with the UseDefaultCredentials of System.Net.Mail.SmtpClient. Sorry for that. 
I saw in another issue that you've commented that MailKit does not implements the UseDefaultCredentials because it uses another API that you do not have access. But I need to use the UseDefaultCredentials.

It works when you send the user and pass, mybad. 

Have a good day!
